### PR TITLE
tower_credential: expect ssh_key_data to be a string instead of path

### LINKED ISSUE #45119: ssh_key_data in tower_credential should not be a file path
##### SUMMARY

`tower_credential` has a parameter named `ssh_key_data`. I want to provide a (vaulted) string there, not a file path. `ssh_key_data` should be renamed to `ssh_key_path/file`.

##### ISSUE TYPE
 - Feature Idea

##### ANSIBLE VERSION

```
ansible 2.7.0b1
```

##### OS / ENVIRONMENT

N/A

##### STEPS TO REPRODUCE

See summary...

I want to be able to store an SSH key in a vaulted variable and not have to deal with files on disk - especially if all the module then does is to read the file as string and send it as string to the tower instance anyways.

##### EXPECTED RESULTS

`ssh_key_data` can contain the data that's in an SSH key file, `ssh_key_file/path` should be the way to refer to an SSH key file instead.

##### ACTUAL RESULTS

A variable named `..._data` is actually referring to a file path.

### PR DESCRIPTION
**Pull request by Meecr0b**
##### SUMMARY
ssh_key_data should be a string filled with the private key
the old behavior can be achieved with a lookup

The `ssh_key_data` variable should behave like in Tower/AWX. It should contain the content of the  ssh_key_data instead of being the path to the ssh_key_file.
The old behavior can be achieved with a lookup.

Fixes #45119

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
tower_credential

##### ANSIBLE VERSION
```
2.8.0
```

### CODE REVIEW

**Comment by MarkusTeufelberger** [comment]
Might need some line(s) in the changelog, but thanks for the quick patch! :+1:

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 1]
```diff
@@ -58,7 +58,9 @@
         - Password for this credential. Use ASK for prompting. secret_key for AWS. api_key for RAX.
     ssh_key_data:
       description:
-        - Path to SSH private key.
+        - SSH private key.
```
Comment (samdoran) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py]: Add a note to use a lookup in order to maintain the previous behavior.

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 2]
```diff
@@ -58,7 +58,9 @@
...
+      required: False
+      default: null
```
Comment (samdoran) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py]: This line is not needed.

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 3]
```diff
@@ -254,13 +256,7 @@ def main():
                 params['team'] = team['id']
 
             if module.params.get('ssh_key_data'):
-                filename = module.params.get('ssh_key_data')
-                if not os.path.exists(filename):
```
Comment (samdoran) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py:258]: Rather than remove this entirely, I would leave this in as a fallback and warn (do not fail) that this behavior will be changed in the future.

File: test/integration/targets/tower_credential/tasks/main.yml [hunk 4]
```diff
@@ -5,6 +5,10 @@
 - name: Generate a local SSH key
   local_action: "shell ssh-keygen -b 2048 -t rsa -f {{ tempdir.stdout }}/id_rsa -q -N 'passphrase'"
 
+- name: Read the generated key
+  local_action: "shell cat {{ tempdir.stdout }}/id_rsa"
```
Comment (samdoran) [inline test/integration/targets/tower_credential/tasks/main.yml]: Use a lookup with `set_fact` rather that `local_action: shell`.

**Review by samdoran — COMMENTED** [review]
Also add an option for `ssh_key_path` and an example of how to use the `file` lookup.

It may be good to add a note to the porting guide for 2.8 as well.

**Comment by Meecr0b** [comment]
Hi @samdoran,
first of all, thx for review. I implemented your comments, except of the additional ``ssh_key_path`` option, because i really don't see the benefits. Can you explain your thoughts?

**Comment by samdoran** [comment]
@Meecr0b I was just thinking you could add it as a new, accurately named option if someone wanted the existing behavior of having the module ingest the key data from a file without having to use a lookup. But it was just a suggestion and if you don't see value in it, then don't worry about adding it.

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 5]
```diff
@@ -254,13 +265,17 @@ def main():
...
-                    module.fail_json(msg='file not found: %s' % filename)
-                if os.path.isdir(filename):
-                    module.fail_json(msg='attempted to read contents of directory: %s' % filename)
-                with open(filename, 'rb') as f:
-                    module.params['ssh_key_data'] = to_text(f.read())
+                data = module.params.get('ssh_key_data')
+                if os.path.exists(data):
+                    module.deprecate(
+                        msg='ssh_key_data should be a string, filled with the private key. Use lookup(\'file\', \'/path/to/file\') instead'
```
Comment (samdoran) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py]: I would reword this message to indicated a path to a file was provided. Something like "ssh_key_data should be a string, not a path to a file".

Also add `version=2.12` the `deprecate()` call.

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 6]
```diff
@@ -254,13 +265,17 @@ def main():
...
+                    )
+                    if os.path.isdir(data):
+                        module.fail_json(msg='attempted to read contents of directory: %s' % data)
```
Comment (samdoran) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py]: I would also add the deprecation warning here in addition to the failure.

**Review by samdoran — CHANGES_REQUESTED** [review]

**Comment by Meecr0b** [comment]
@samdoran thx for explanation. In this case, i would prefer the ansible native solution, instead of supporting custom python code.

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py (hunk 6 above)
Comment (Meecr0b) [reply lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py]: The failure is just an additional error, the depraction warning will be returned in any case.
Or did i misunderstand you?

Comment (samdoran) [reply lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py]: You're correct. I misread that this `if` statement is indented. Looks good!

**Review by samdoran — APPROVED** [review]

**Comment by samdoran** [comment]
Please create a changelog fragment. See [fragments](https://github.com/ansible/ansible/blob/stable-2.6/changelogs/fragments/user-properly-unexpire-new-user.yaml) for examples.

**Comment by Meecr0b** [comment]
Thx for example, [done](https://github.com/Meecr0b/ansible/blob/devel/changelogs/fragments/tower_credential_ssh_key_data.yaml).

**Comment by acozine** [comment]
Thanks @Meecr0b for contributing! I wordsmithed the documentation a bit. PR looks good to me.

**Comment by mattclay** [comment]
This change appears to be causing the tower_project integration test to fail:

https://app.shippable.com/github/ansible/ansible/runs/83068/90/tests
